src/AzurePipelinesToGitHubActionsConverter.Core/Conversion/VariablesProcessing.cs

```diff
@@ -314,5 +398,91 @@ private List<string> FindPipelineParametersInString(string text)
 
             return list;
         }
+
+        public MatchCollection FindPipelineVariables(string input)
+        {
+            // match anything in ${}, ${{}}, $(), $[], but NOT $var
+            //  Allowed ADO var chars here: https://docs.microsoft.com/en-us/azure/devops/pipelines/process/variables?view=azure-devops&tabs=yaml%2Cbatch#variable-characters
+            var varPattern = @"\$(?:\{|\{|\(|\[|\{\{)([^\r\n(){}\[\]$]+)(?:\}\}|\}|\]|\})?(?:\}\}|\]|\)|\}|\})";
+
+            return Regex.Matches(input, varPattern);
+        }
+
+        public MatchCollection FindPipelineVariable(string input, string var)
+        {
+            // match "var" in ${}, ${{}}, $(), $[], but NOT $var
+            var varPattern = string.Format(@"\$(?:\{|\{|\(|\[|\{\{)({0})(?:\}\}|\}|\]|\})?(?:\}\}|\]|\)|\}|\})", var);
+
+            return Regex.Matches(input, varPattern, RegexOptions.IgnoreCase);
+        }
+
+        public string ProcessVariableConversions(string yaml, string matrixVariableName = null)
+        {
+            // Replace variables with the format "${{ [prefix.]MyVar }}"
+            var matches = FindPipelineVariables(yaml);
+
+            foreach (Match match in matches)
+            {
+                /// actual var name will be returned in grouping 1
+                var varName = match.Groups[1].Value;
+
+                // Only do replacement if this is one of the vars we've identified... stops false positives like job.status
+                if (VariableList.Contains(varName))
+                {
+                    if (varName != matrixVariableName)
+                    {
+                        yaml = yaml.Replace(match.Value, $"${{{{ env.{varName} }}}}");
+                    }
+                    else // matrix var
+                    {
+                        yaml = yaml.Replace(match.Value, $"${{{{ matrix.{varName} }}}}");
+                    }
+                }
+                else // see if this matches a system var we know how to replace
+                {
+                    var systemVar = SystemVarMapping.FirstOrDefault(v => v.Key.ToLower() == varName.ToLower());
+
+                    if (systemVar.Value != null)
+                    {
+                        yaml = yaml.Replace(match.Value, $"${{{{ {systemVar.Value} }}}}");
+                    }
+                    else // not a var we've defined or a known system var, but we'll convert the syntax
+                    {
+                        yaml = yaml.Replace(match.Value, $"${{{{ env.{varName} }}}}");
+                    }
+                }
+            }
+
+            yaml = yaml.ReplaceAnyCase("${{ env.rev:r }}", "${ GITHUB_RUN_NUMBER }"); // need to verify, moved over from older code; prefer prettier [github.] context usage
```

Comment:
Does this still need to be verified?  If it's been verified please remove the comment.

Reply:
Yes - this conversion was already in place in the old code... we don't seem to hit this case in the MC pipelines so I feel it does need to be verified at some point as part of the system var conversion review we discussed.

Reply:
I think it depends on the context.  When used in expressions, I think it should be ${{ github.run_number }} but it should be ${ GITHUB_RUN_NUMBER } when used in a script.  What's the context here?  It looks like it's an expression.

Reply:
I used the `github` context in all of the replacements I've added, and I think that is likely appropriate here (and looks way better). The `github` context is also available via expressions in script steps - I've verified this. External scripts likely need the env-style syntax if they want to leverage those values, but of course wouldn't be included in the use case here of replacing ADO-based values.

I can go ahead and just change it to use the `github` context if we agree that's appropriate, but I just didn't want to change things included in the old code w/o a use case to test it with - hence the comment and bringing it over as is.